src/services/rfqm_service.ts

```diff
@@ -0,0 +1,138 @@
+    constructor(private readonly _quoteRequestor: QuoteRequestor) {}
+
+    /**
+     * Fetch the best indicative quote available
+     */
+    public async fetchIndicativeQuoteAsync(params: FetchIndicativeQuoteParams): Promise<FetchIndicativeQuoteResponse> {
+        const { sellAmount, buyAmount, sellToken, buyToken, apiKey } = params;
+
+        // Map params to the terminology of Quote Requestor
+        const isSelling = sellAmount !== undefined;
+        const marketOperation = isSelling ? MarketOperation.Sell : MarketOperation.Buy;
+        const assetFillAmount = isSelling ? sellAmount! : buyAmount!;
+        const makerToken = buyToken;
+        const takerToken = sellToken;
+
+        // Fetch quotes
+        const opts = {
+            takerAddress: NULL_ADDRESS,
+            txOrigin: NULL_ADDRESS, // TODO - set to worker registry
+            apiKey,
+            intentOnFilling: false, // TODO - safe to hardcode?
+            isIndicative: true,
+            makerEndpointMaxResponseTimeMs: RFQT_REQUEST_MAX_RESPONSE_MS,
+            nativeExclusivelyRFQ: true,
+            altRfqAssetOfferings: {}, // TODO - set this to the alt rfq asset offerings
+            isLastLook: true,
+        };
+        const indicativeQuotes = await this._quoteRequestor.requestRfqmIndicativeQuotesAsync(
+            makerToken,
+            takerToken,
+            assetFillAmount,
+            marketOperation,
+            undefined,
+            opts,
+        );
+
+        // Filter out quotes that:
+        // - are for the wrong pair
+        // - cannot fill 100 % of the requested amount
+        //
+        // And sort by best price
+        const sortedQuotes = indicativeQuotes
+            .filter((q) => q.takerToken === takerToken && q.makerToken === makerToken)
+            .filter((q) => q.takerAmount.gte(assetFillAmount))
+            .sort((a, b) => {
+                const aPrice = a.makerAmount.div(a.takerAmount);
+                const bPrice = b.makerAmount.div(b.takerAmount);
+                return bPrice.minus(aPrice).toNumber();
+            });
+
+        // No quotes found
+        if (sortedQuotes.length === 0) {
+            return Promise.reject(new Error('No valid quotes'));
+        }
+
+        const bestQuote = sortedQuotes[0];
+
+        // Prepare response
+        // TODO: handle decimals properly in price
+        return {
+            buyAmount: bestQuote.makerAmount,
+            buyTokenAddress: bestQuote.makerToken,
+            sellAmount: bestQuote.takerAmount,
+            sellTokenAddress: bestQuote.takerToken,
+            price: isSelling
+                ? bestQuote.makerAmount.div(bestQuote.takerAmount)
+                : bestQuote.takerAmount.div(bestQuote.makerAmount),
+        };
+    }
+
+    /**
+     * Fetch the best firm quote available, as a signable metatransaction
+     */
+    public async fetchFirmQuoteAsync(params: FetchFirmQuoteParams): Promise<FetchFirmQuoteResponse> {
```

Comment:
I noticed the function implementation is pretty much identical. Do you anticipate them diverging, or will most of the business logic be handled by QuoteRequestor? If the latter, would suggest DRYing up this code. 

e.g. 
```
private async _fetchQuoteAsync(params: RfqmQuoteParams, isIndicative: boolean): { 
    // same implementation, just pass isIndicative directly through to opts
}
```

And `fetchFirmQuote` can do additional work to create a metatx.

Reply:
Good idea - will give this a shot

Reply:
Gave it a shot - but after wrestling with some of our types (so many union types), will punt on this for now.

I'm going to remove fetching firm quotes at the moment, as it is far from being complete